src/common/ui/codemirror-js-mixed/javascript-mixed.js

```diff
@@ -0,0 +1,756 @@
+      return stream.pos - stream.start;
+    }
+
+    function prepReparseStringTemplateInLocalMode(modeToUse, stream, state,
+      hasBeginBacktick = true) {
+      dbg(`Entering local ${modeToUse.name} mode...`);
+      if (hasBeginBacktick) {
+        // spit out beginning backtick as a token, and leave the rest of the text for local mode parsing
+        stream.backUp(tokenLength(stream) - 1);
+      } else {
+        stream.backUp(tokenLength(stream));
+      }
+
+      // workaround needed for 1-line string template,
+      // to ensure the ending backtick is parsed correctly.
+      forceJsModeToQuasi(stream, state.jsState);
+
+      // switch to local mode for subsequent text
+      state.localMode = modeToUse;
+      state.localState = CodeMirror.startState(state.localMode);
+      state.inJsExprInStringTemplate = false;
+      state.jsExprDepthInStringTemplate = 0;
+    }
+
+    function tokenLastChar(stream) {
+      return stream.string.charAt(stream.pos - 1) || undefined;
+    }
+
+    function isEndBacktick(stream, state) {
+      // check it hits ending backtick for string template,
+      // ignoring the backticks that appear inside a JS expression.
+      return !state.inJsExprInStringTemplate && stream.peek() === '`'
+        && tokenLastChar(stream) !== '\\'; // ensure it is not an escaped backtick (doesn't count)
+    }
+
+    function exitLocalModeWithEndBacktick(stream, state) {
+      dbg('Exiting local html/css mode...');
+      // parse the ending JS string template backtick in js mode
+      return jsMode.token(stream, state.jsState);
+    }
+
+    // Local mode-specific helpers to handle js expression in string template
+    function curModeNameAndStateOfHtmlmixed(htmlmixedState) {
+      if (!htmlmixedState.localMode) {
+        return ['html', htmlmixedState.htmlState];
+      }
+      // else return local css / javascript name, state
+      return [htmlmixedState.localMode.name, htmlmixedState.localState];
+    }
+
+    /**
+     * @return the position of '${' relative to the
+     *         current token start position, i.e., stream.start; -1 otherwise.
+     */
+    htmlmixedMode.indexOfJsExprStartInStream = function indexOfJsExprStartInStream(stream, state) {
+      const [modeName, modeState] = curModeNameAndStateOfHtmlmixed(state.localState);
+      switch (modeName) {
+      case 'html':
+        return indexOfJsExprStartInString(stream.current());
+      case 'css':
+        return cssMode.indexOfJsExprStartInStream(stream, modeState);
+      case 'javascript':
+        return -1; // let js mode handle ${ natively
+      default:
+        console.error('htmlmixedMode.indexOfJsExprStartInStream() - unrecognized mode:', modeName, modeState);
+      }
+      return -1; // should never reach here
+    };
+
+    // eslint-disable-next-line max-len, no-unused-vars
+    htmlmixedMode.ensureProperLocalModeStatePostJsExpr = function ensureProperLocalModeStatePostJsExpr(stream, state, style) {
```

Comment:
You're not using `this` so no need for `function`. I guess you wanted it to have a name which can be supplanted differently I think:

```js
Object.assign(htmlmixedMode, {
  ensureProperLocalModeStatePostJsExpr(stream, state, style) {
    ........
  },
});
```

Or maybe even like this:

```js
const htmlmixedMode = CodeMirror.getMode(config, {
  name: 'htmlmixed',
  ensureProperLocalModeStatePostJsExpr(stream, state, style) {
    ........
  },
});
```

Bonus: there'll be no need for eslint overrides.

Reply:
Other than stylistic preferences, what's the reason the `function` is not preferrable?

Anyway, I changed to `Object.assign`, which also let me group related functions in a block, and keep the logic close to where they are called.

Reply:
The problem I wanted to solve is the use of eslint overrides, duplication, and excessive length.